src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+            } else {
+                ret = LINE_XOXO; // '|'
+            }
+        }
+    }
+
+    return ret;
+}
+
+void auto_bordered_window::local_draw()
+{
+    ui_window::local_draw();
+
+    auto win = get_win(); // never null
+    for( unsigned int x = 0; x < get_rect().size_x; x++ ){
+        for( unsigned int y = 0; y < get_rect().size_y; y++ ) {
+            if( is_uncovered( x, y ) ) {
+                mvwputch( win, y, x, border_color, get_border_char( x, y ) );
+            }
+        }
+    }
+}
+
+ui_vertical_list::ui_vertical_list( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor ) : ui_element( size_x, size_y, x, y, anchor )
+{
+}
+
+void ui_vertical_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int start_line = window_scroll;
+    unsigned int end_line = start_line + get_rect().size_y;
+
+    unsigned int available_space = get_rect().size_x - 2; // 2 for scroll bar and spacer
+
+    for( unsigned int line = start_line; line < end_line && line < text.size(); line++ ) {
+        auto txt = text[line];
+        if( txt.size() > available_space ) {
+            txt = txt.substr( 0, available_space );
+        }
+        if( scroll == line ) {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, hilite(text_color), txt.c_str() );
+        } else {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, text_color, txt.c_str() );
+        }
+    }
+
+    draw_scrollbar( win, scroll, get_rect().size_y, text.size(), get_ay(), get_ax(), bar_color, false );
+}
+
+void ui_vertical_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+}
+
+void ui_vertical_list::scroll_up()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+
+    if( scroll == text.size() - 1 ) {
+        window_scroll = scroll - get_rect().size_y + 1;
+    } else if( scroll < window_scroll ) {
+        window_scroll--;
+    }
+}
+
+void ui_vertical_list::scroll_down()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+
+    if( scroll > get_rect().size_y + window_scroll - 1 ) {
+        window_scroll++;
+    } else if( scroll == 0 ) {
+        window_scroll = 0;
+    }
+}
+
+const std::string &ui_vertical_list::current() const
+{
+    return text[scroll];
+}
+
+ui_horizontal_list::ui_horizontal_list( int x, int y, ui_anchor anchor ) : ui_element( 0, 1, x, y, anchor )
+{
+}
+
+void ui_horizontal_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int x_offset = 1 + get_ax();
+
+    for( unsigned int i = 0; i < text.size(); i++ ) {
+        x_offset += draw_subtab( win, get_ax() + x_offset, get_ay(), text[i], i == scroll ) + 2;
+    }
+}
+
+void ui_horizontal_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+    size_t text_length = 1;
+
+    for( const auto &str : text ) {
+        text_length = utf8_width( str ) + 1;
```

Comment:
`text_length` is simply overridden in each iteration? Why not just take the length of the last string: `text_length = utf8_width( text.back() ) + 1`